### PR TITLE
Retry some operations

### PR DESCRIPTION
**Pull request by mjtv**
Allows recovery from transient errors such as connection refused during a broker restart
Fixes #226

### CODE REVIEW

File: examples/main.tf [hunk 1]
```diff
@@ -36,6 +36,13 @@ resource "kafka_topic" "syslog" {
     "retention.ms" = "86400000"
   }
 
+  timeouts {
+    create = "8m"
+    read   = "8m"
+    update = "8m"
+    delete = "8m"
+  }
```
Comment (mjtv) [inline examples/main.tf:44]: Standard terraform CRUD timeouts
https://www.terraform.io/plugin/sdkv2/resources/retries-and-customizable-timeouts

File: kafka/client.go [hunk 2]
```diff
@@ -25,6 +26,7 @@ type Client struct {
 	kafkaConfig   *sarama.Config
 	config        *Config
 	supportedAPIs map[int]int
+	topicsMutex   sync.Mutex
```
Comment (mjtv) [inline kafka/client.go:29]: Encountered a "concurrent map writes" panic during testing
map is not thread safe
Pre-existing issue

File: kafka/data_source_kafka_topic.go [hunk 3]
```diff
@@ -36,23 +39,24 @@ func kafkaTopicDataSource() *schema.Resource {
...
-func dataSourceTopicRead(d *schema.ResourceData, meta interface{}) error {
+func dataSourceTopicRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	// Unlike the resource topicRead, there is no pre-existing ID. We must use the 'name' to look up the resource.
 	// See https://learn.hashicorp.com/tutorials/terraform/provider-create?in=terraform/providers#implement-read
 	name := d.Get("name").(string)
 
-	client := meta.(*LazyClient)
-	topic, err := client.ReadTopic(name, true)
+	client := meta.(*RetryLazyClient)
+	timeout := 10 * time.Minute
```
Comment (mjtv) [inline kafka/data_source_kafka_topic.go:48]: Terraform only provides timeouts to resources not datasources so using a hard coded 10 minute timeout

File: kafka/kafka_acls.go [hunk 4]
```diff
@@ -424,7 +424,7 @@ func (c *Client) ListACLs() ([]*sarama.ResourceAcls, error) {
 		if err != nil {
 			return nil, err
 		}
-		
+
```
Comment (mjtv) [inline kafka/kafka_acls.go:427]: Change is a result of using "go fmt"

File: kafka/lazy_client.go [hunk 5]
```diff
@@ -10,19 +10,26 @@ import (
 )
 
 type LazyClient struct {
-	once    sync.Once
+	mutex   sync.Mutex
```
Comment (mjtv) [inline kafka/lazy_client.go:13]: Switch from once to a mutex to support retries

File: kafka/resource_kafka_acl.go [hunk 6]
```diff
@@ -60,15 +61,22 @@ func kafkaACLResource() *schema.Resource {
 				ForceNew: true,
 			},
 		},
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(10 * time.Minute),
+			Read:   schema.DefaultTimeout(10 * time.Minute),
+			Update: schema.DefaultTimeout(10 * time.Minute),
+			Delete: schema.DefaultTimeout(10 * time.Minute),
+		},
```
Comment (mjtv) [inline kafka/resource_kafka_acl.go:69]: Standard terraform CRUD timeouts
https://www.terraform.io/plugin/sdkv2/resources/retries-and-customizable-timeouts

File: kafka/resource_kafka_topic.go [hunk 7]
```diff
@@ -50,23 +50,31 @@ func kafkaTopicResource() *schema.Resource {
 				Elem:        schema.TypeString,
 			},
 		},
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(10 * time.Minute),
+			Read:   schema.DefaultTimeout(10 * time.Minute),
+			Update: schema.DefaultTimeout(10 * time.Minute),
+			Delete: schema.DefaultTimeout(10 * time.Minute),
+		},
```
Comment (mjtv) [inline kafka/resource_kafka_topic.go:58]: Standard terraform CRUD timeouts
https://www.terraform.io/plugin/sdkv2/resources/retries-and-customizable-timeouts

File: kafka/resource_kafka_acl.go [hunk 8]
```diff
@@ -60,15 +61,22 @@ func kafkaACLResource() *schema.Resource {
...
 	}
 }
 
 func aclCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	c := meta.(*LazyClient)
+	timeout := d.Timeout(schema.TimeoutCreate) - time.Minute
```
Comment (mjtv) [inline kafka/resource_kafka_acl.go:74]: Subtract time.Minute as suggested in the Terraform Plugin SDKv2 docs
https://www.terraform.io/plugin/sdkv2/resources/retries-and-customizable-timeouts#retry

File: kafka/retry_lazy_client.go [hunk 9]
```diff
@@ -0,0 +1,346 @@
+package kafka
```
Comment (mjtv) [inline kafka/retry_lazy_client.go:1]: Allow recovery from transient errors such as connection refused during a broker restart

RetryLazyClient wraps the LazyClient with retries and timeouts similar to how LazyClient wraps the Client with lazy instantiation

Uses the Terraform Plugin SDKv2 retry mechanism
https://www.terraform.io/plugin/sdkv2/resources/retries-and-customizable-timeouts
https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource

**Comment by mjtv** [comment]
Reminder: This pull request is for code review only. We will rebase the tivo-build branch on top of this retry-2 branch.

File: kafka/lazy_client.go [hunk 10]
```diff
@@ -10,19 +10,26 @@ import (
...
 	initErr error
 	inner   *Client
 	Config  *Config
 }
 
 func (c *LazyClient) init() error {
-	var err error
+	func(c *LazyClient) {
```
Comment (benoittoulme) [inline kafka/lazy_client.go:20]: why declaring a function in this function to just call it?

Comment (benoittoulme) [reply kafka/lazy_client.go:20]: You are setting c.initErr in this function but using the value below. There is a possible race condition if the value is set from exiting this function to reading the value below.

Comment (benoittoulme) [reply kafka/lazy_client.go:20]: I think I would lock the mutex for the whole duration of the init() call.

Comment (mjtv) [reply kafka/lazy_client.go:20]: Good catch. I see how a race condition could happen and a transient error that was encountered might not be fully logged. Performance isn't critical here so I will change it.

**Review by benoittoulme — APPROVED** [review]

**Pull request closed without merging**